### PR TITLE
Make a RootDescription not depend on the root Package

### PR DESCRIPTION
**Pull request by sigurdm**
This is refactor that is extracted from https://github.com/dart-lang/pub/pull/4060 . It will allow a RootDescription to be created before the pubspec is loaded.

This removes the concept of an "in-memory package". These packages were always a modified version of a pubspec, being used for trying out a resolution, and they might as well be associated with the directory of the original pubspec.

### CODE REVIEW

**Review by jonasfj — APPROVED** [review]
LGTM; but it might be worth having @szakarias have a look too.

File: lib/src/package.dart [hunk 1]
```diff
@@ -134,21 +119,14 @@ class Package {
...
-
-  /// Constructs a package with the given pubspec.
+  /// Creates a package with [pubspec] associated with [dir].
   ///
-  /// The package will have no directory associated with it.
-  Package.inMemory(this.pubspec) : _dir = null;
-
-  /// Creates a package with [pubspec] located at [dir].
-  Package(this.pubspec, String this._dir);
+  /// For temporary resolution attempts `pubspec` does not have to correspond
```
Comment (szakarias) [inline lib/src/package.dart:127]: [pubspec] ?

File: lib/src/source/root.dart [hunk 2]
```diff
@@ -5,12 +5,21 @@
...
 import '../pubspec.dart';
 import '../source.dart';
 import '../system_cache.dart';
 
+/// A root package is the root of a package dependency graph that seeds a
+/// version resolution.
+///
+/// There is no explicit way to depend on a root package.
+///
+/// A root package is the only package for whom dev_dependencies are taken into
```
Comment (szakarias) [inline lib/src/source/root.dart:18]: for which?

**Review by szakarias — APPROVED** [review]
Nice one.

File: lib/src/source/root.dart (hunk 2 above)
Comment (sigurdm) [reply lib/src/source/root.dart]: Done

File: lib/src/source/root.dart [hunk 3]
```diff
@@ -5,12 +5,21 @@
...
+/// account.
+///
+/// A root package is the only package for whom dependency_overrides are taken
```
[the remark this answers is not in the document]
Comment (sigurdm) [reply lib/src/source/root.dart]: Done

**Pull request merged**